File: binartSearch/cpp/34.find-first-and-last-position-of-element-in-sorted-array.cpp
/*
 * @lc app=leetcode id=34 lang=cpp
 *
 * [34] Find First and Last Position of Element in Sorted Array
 */

// @lc code=start
#include <vector>
using namespace std;

class Solution {
public:
    int findLeft(const vector<int> &nums, const int &target) {
        int left = 0, right = nums.size() - 1;
        int index = -1;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                index = mid;
                right = mid - 1;
            } else if (nums[mid] > target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return index;
    }

    int findRight(const vector<int> &nums, const int &target) {
        int left = 0, right = nums.size() - 1;
        int index = -1;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                index = mid;
                left = mid + 1;
            } else if (nums[mid] > target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return index;
    }

    vector<int> searchRange(vector<int>& nums, int target) {
        return vector<int> {findLeft(nums, target), findRight(nums, target)};
    }
};
// @lc code=end

File: binarySearch/python/34.find-first-and-last-position-of-element-in-sorted-array.py
#
# @lc app=leetcode id=34 lang=python3
#
# [34] Find First and Last Position of Element in Sorted Array
#

# @lc code=start
class Solution:
    def findLeft(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1
        index = -1
        while (left <= right):
            mid = int((left + right) / 2)
            mid_value = nums[mid]
            if mid_value == target:
                right = mid - 1
                index = mid
            elif mid_value < target:
                left = mid + 1
            else:
                right = mid - 1
        
        return index

    def findRight(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1
        index = -1
        while (left <= right):
            mid = int((left + right) / 2)
            mid_value = nums[mid]
            if mid_value == target:
                left = mid + 1
                index = mid
            elif mid_value < target:
                left = mid + 1
            else:
                right = mid - 1
        
        return index
                

    def searchRange(self, nums: List[int], target: int) -> List[int]:
        return [self.findLeft(nums, target), self.findRight(nums, target)]
# @lc code=end

